qutip/core/data/convert.pyx

```diff
@@ -0,0 +1,285 @@
+#cython: language_level=3
+
+"""
+The conversion machinery between different data-layer types, and creation
+routines from arbitrary data.  The classes `_to` and `_create` are not intended
+to be exported names, but are the backing machinery of the functions `data.to`
+and `data.create`, which are built up as the last objects in the `__init__.py`
+initialisation of the `data` module.
+"""
+
+# This module is compiled by Cython because it's the core of the entire
+# dispatch table, and having it compiled to a C extension saves about 1µs per
+# call.  This is not much at all, and there's very little which benefits from
+# Cython compiliation, but such core functionality is called millions of times
+# even in a simple interactive QuTiP session, and it all adds up.
```

Comment:
Didn't you talk about reducing our reliance on cython when proposing your project...

Reply:
This is at compile-time only, not at run-time.  The user doesn't need to have Cython installed.  I thought we'd talked about reducing Cython usage only in the context of string coefficients?

At any rate, the entire data layer is implemented in Cython, and there's no escaping that while keeping up the speed gains.  New data-types don't need to be specified using Cython at all, so there's no run-time requirement on it.

Reply:
As in, the user never has to recompile this object - you can add new data types at runtime from within Python space, no problem at all.

Reply:
I have no issue with it. I was commenting on how it seems you got to like cython during the last few months.

Reply:
Oh yeah!  Cython's definitely got a very steep initial learning curve, and its documentation is a bit lacking in places, but I'm definitely a big fan for the really highly used components.  One thing I am careful about is implementing too much in Cython - it makes it much harder for downstream users to monkeypatch our functionality to suit their own needs.